src/memory/AmdFlash.cc

```diff
@@ -35,152 +41,125 @@ AmdFlash::AmdFlash(const std::string& name, const ValidatedChip& validatedChip,
-	}
-	if (readOnlySize) {
-		// If some part of the flash is read-only we require a ROM
-		// constructor parameter.
-		assert(rom);
+	for (const auto [sector, writeProtect] : enumerate(writeProtectSectors)) {
+		sectors[sector].writeProtect = writeProtect;
 	}
 
-	const auto* romTag = config.getXML()->findChild("rom");
-	bool initialContentSpecified = romTag && romTag->findChild("sha1");
-
-	// check whether the loaded SRAM is empty, whilst initial content was specified
-	if (!rom && loaded && initialContentSpecified && sramEmpty(*ram)) {
-		config.getCliComm().printInfo(
-			"This flash device (", config.getHardwareConfig().getName(),
-			") has initial content specified, but this content "
-			"was not loaded, because there was already content found "
-			"and loaded from persistent storage. However, this "
-			"content is blank (it was probably created automatically "
-			"when the specified initial content could not be loaded "
-			"when this device was used for the first time). If you "
-			"still wish to load the specified initial content, "
-			"please remove the blank persistent storage file: ",
-			ram->getLoadedFilename());
-	}
+	File file;
+	try {
+		const auto& filename = config.getChildData("sramname");
+		const std::string path = config.getFileContext().resolveSavePaths(filename);
+		const std::string metaPath = path + ".meta";
+
+		if (FileOperations::exists(metaPath)) {
+			try {
+				XmlInputArchive in(metaPath);
+				in.serialize("sectors", sectors);
+			} catch (MSXException& e) {
+				config.getCliComm().printWarning("Couldn't load Flash metadata ",
+					filename, ".meta (", e.getMessage(), ").");
+			}
+		} else { // back compat
+			for (Sector& sector : sectors) {
+				sector.modified = !sector.writeProtect;
+			}
+		}
 
-	std::unique_ptr<Rom> rom_;
-	if (!rom && !loaded) {
-		// If we don't have a ROM constructor parameter and there was
-		// no sram content loaded (= previous persistent flash
-		// content), then try to load some initial content. This
-		// represents the original content of the flash when the device
-		// ships. This ROM is optional, if it's not found, then the
-		// initial flash content is all 0xFF.
 		try {
-			rom_ = std::make_unique<Rom>(
-				"", "", // dummy name and description
-				config);
-			rom = rom_.get();
-			config.getCliComm().printInfo(
-				"Loaded initial content for flash ROM from ",
-				rom->getFilename());
-		} catch (MSXException& e) {
-			// ignore error
-			assert(rom == nullptr); // 'rom' remains nullptr
-			// only if an actual sha1sum was given, tell the user we failed to use it
-			if (initialContentSpecified) {
-				config.getCliComm().printWarning(
-					"Could not load specified initial content "
-					"for flash ROM: ", e.getMessage());
-			}
+			file = File(path, File::OpenMode::LOAD_PERSISTENT);
+		} catch (FileException& e) {
+			config.getCliComm().printWarning("Couldn't load Flash data ",
+				filename, " (", e.getMessage(), ").");
```

Comment:
The call to config.getFileContext().resolveSavePaths(filename) will throw if the file doesn’t exist, which the catch on line 86 below it will receive, so it behaves the same way as in the previous patch.

Though aside, not a big fan of how these resolve functions communicate failure by throwing, it leads to confusions like this because you can’t tell by just looking at it that the function can fail (and with which exception). Would perhaps be nicer if they returned an optional or variant.

Reply:
Aha, I understand now.
I agree with you. In _this_ case a non-throwing version would be much cleaner.

In the majority of the cases where we use `resolve()` it is a hard error if the file cannot be found. And then it's nice that the error handling (via an exception) is automatic. But here we **do** want to handle that case. I wouldn't mind offering two versions: a `resolve()` like we have now, and an extra `tryResolve()` method which returns an `std::optional`. (Suggestions for better names are welcome ofcourse).

In the XMLElement class we have something similar:  the `getChild()` (and variations) methods throw an exception (it's a hard error if the tag is missing). But the `findChild()` methods may return a `nullptr`, these are used to query optionally present tags.

(And in our defense, this code is +20 years old, at the time `std::optional` and `std::variant` didn't exist yet. But that doesn't mean we shouldn't try to improve it now).